Refactor typing/env to separate the filesystem-related logic

gasche opened the pull request:
This PR, which builds on top of #2227, is again only a refactoring PR: it should not affect the semantics of the compiler (there are two small changes in behavior that I mention below; the rest should be strictly identical, only moving code and data around).

The PR came as I was trying to understand the code of Env that is related to compilation units. `typing/env.ml` has more than 2K lines, and about 300 of them are related to compilation units (reading and writing `cmi` files, consistency checking, various caches, etc.), but they are blended in with the rest in a way that makes it difficult (for me) to follow the logic of this part of the code.

The present PR separates this persistent-files logic to a separate module, Persistent_env, that Env then uses internally. The last commit does the actual file split, all other preceding commits are small changes that were necessary for it, either because they abstract away certain details of the Env module that are affected by the split, or because they disentangle the mess of dependencies within the module that makes the split difficult -- it took me 3 attempts to discover everything that could create a cyclic dependency.

The change might be useful to the Merlin people (cc @trefis, @let-def), as Merlin handles cmi files in a different way from the upstream type-checker.

Now on the two small behavior changes:

1. The `avoid external uses of {add_import,crc_units}` commit merges together three loops that are almost identical, but have a difference in the way `Env.add_import` is called or not on transitive dependencies that have no CRC. This specific change should be reviewed by @lpw25. (Understanding the mess of direct/indirect and weak/strong and transparent/opaque dependencies is my next goal.) Making this merge only makes the overall code cleaner, it's easy to separate into two functions if it turns out to be incorrect.

2. The `remove the current unit name from some exceptions` commit removes some information from Persistent_env exception that was constructed, at raising sites, by using internal state from Env (the name of the current compilation unit). This changes the error messages slightly (see the commit message for an example), but I believe the change is harmless. On the other hand, trying to preserve the previous behavior exactly would be doable, but make the code more complex (I could register the exception-handler in `Env` instead of `Persistent_env`, to have the information available at printing-time, but the irregular structure would be a bit vexing).

gasche committed:
env refactoring: isolate unit_name reference

gasche committed:
env refactoring: factorize clear/reset logic

gasche committed:
env refactoring: avoid external uses of {add_import,crc_units}

There is a small change of behavior in this patch due to a different
handling of weak dependencies (those with crco=None); in
Env.check_consistency, only non-weak dependencies would get
[Env.add_import] called, while the `toplevel/` implementations would
also call [Env.add_import] on weak dependencies. After this patch, we
systematically call [add_import] only on non-weak dependencies, even
in `toplevel/`.

([Gabriel:] As far as I can see, the use of [add_import] in the
toplevel never leads to a use of [Env.imports()] for producing
a dependency list, as the toplevel does not produce cmi/cmo files; are
they just no-ops?)

gasche committed:
env refactoring: aggregate the state for persistent structures into one record

gasche committed:
env refactoring: separate consistency from typing logic in persistent_env

gasche committed:
env refactoring: move EnvLazy to utils/misc.ml

gasche committed:
env refactoring: separate the cmi/crc handling into a persistent_env module

Persistent_env is a new module that handles the relation between the
type-checking state and the "persistent" typing information laying in
.cmi files on the filesystem. In particular, it handles the collection
and production of CRC information for the .cmi files being read and
written to the filesystem; the using modules (in our case, only Env)
are in charge of turning the cmi files into higher-level information
(components and signatures).

Persistent_env exposes a type `'a t` of a persistent environment,
which acts as a mutable store of `'a` values. There is no global state
in the module itself: while Env (and thus the OCaml type-checker) uses
a single global persistent environment, it should be possible to
create several independent environments to represent, for example,
several independent type-checking sessions.

gasche committed:
env refactoring: move code around (no change)

gasche committed:
env refactoring: remove the current unit name from some exceptions

This small change of behavior simplifies the internal plumbing of env
by avoiding the need to passes the 'current_unit_name' state to
cmi-checking exceptions -- it allows to separate the cmi/crc logic to
a separate module in a future commit.

We believe that the change does not actually reduce error message
clarity, as the name of the offending unit appears in the location
filename anyway (see how these exceptions are handled by
Location.error_of_printer_file in the error printer).

Before:

    File "a.mli", line 1:
    Error: Unit A imports from B, which uses recursive types.
           The compilation flag -rectypes is required

After:

    File "a.mli", line 1:
    Error: Invalid import of B, which uses recursive types.
           The compilation flag -rectypes is required

gasche committed:
Changes entry

gasche commented:
I just handled a somewhat-painful rebase on top of #2041 (now merged in trunk). Nothing much to declare (I wasn't sure whether `persistent_structures_of_dir` should stay in Env or move in Persistent_env, I left it in Env for now). @diml, I think you might be interested in (reviewing?) the present PR: I think that some of the code structure changes I made reasonate with some of the parts you touched; maybe the split in this PR would have made your work easier.

gasche commented:
P.S.: I made sure that the PR is split in several clean commits, and it should be easiest to review commit-by-commit.

ghost commented:
@gasche sure, I'm having a look

diml commented on typing/persistent_env.ml:
```diff
@@ -0,0 +1,344 @@
+module String = Misc.Stdlib.String
+
+type 'a t = {
+  persistent_structures : (string, (pers_struct * 'a) option) Hashtbl.t;
```
Maybe we should use a specific datatype here. I didn't immediately remember what `None` meant. For instance:

```ocaml
type 'a pers_struct_info =
  | Not_found
  | Found of { pers_struct : pers_struct; data : 'a }

type 'a t = {
  persistent_structures : (string, 'a pers_struct_info) Hashtbl.t;
...
}
```

gasche replied:
Indeed, thanks! I used `Found` and `Missing` to avoid ambiguities with the `Not_found` exception (and also `missing` is already used in the API), and unnamed parameters instead of a record because I don't know of a very good name for `data`.

gasche commented:
@diml Thanks! I'll adress your comments shortly. Do you have an opnion on the interest of the refactoring -- do you agree that it makes the final code nicer, or does it seems roughly equivalent to you?

gasche committed:
env refactoring: move current_unit in a submodule

(suggested by Jérémie Dimino)

gasche committed:
env refactoring: use a custom ADT for missing pers_struct entries

The hope is that a tailor-made algebraic datatype is more readable /
less confusing than using ('a option) directly -- one may confuse
getting None when looking in a table with the Not_found case.

(Suggested by Jérémie Dimino)

gasche commented:
(I pushed two additional commits to implement the suggestions of @diml)

gasche commented:
I forgot to say: I grepped the code of `utop` to see how `Env` is used there (in particular I was worried about uses of `add_import`, etc.), and I believe that it would not require any code change with this PR. (Otherwise I would change the PR, I guess.)

gasche committed:
update the code in tests/self-contained-toplevel

The hope is that the (env => persistent_env) refactoring does not
break reasonable user code; the fact that this test had to be updated
is a bad sign. On the other hand, we believe that utop is unaffected
by the change, which suggests that real-world toplevel are less likely
to be affected.

gasche commented on typing/env.ml:
```diff
@@ -677,80 +555,27 @@ let rec print_address ppf = function
-
-let add_import s =
-  imported_units := String.Set.add s !imported_units
+let set_unit_name name =
```
Done. (It is not super-important to protect this reference, so we could revert this change. Initially I thought of moving the reference into Persistent_env, which is why I needed this preparatory work.)

ghost commented:
I agree that it makes the final code look nicer. I find `Env` a bit hard to digest, so I welcome splitting it into smaller individual parts.

ghost commented:
BTW, in utop we use cppo to be compatible with several versions of the compiler, so we can adapt to breaking changes in the compiler API.

gasche commented:
@diml (and I guess @trefis?): merging this now could create rebase pains for #2235, so I propose to delay a bit here until you know what to do there.

trefis commented:
I haven't looked at the code in this PR in details yet but:
- I don't think this should be too painful for #2235, but it might be for #2231 (which I think should also make it into 4.08)
- I'm surprised by the copyright header on `persistent_env.ml[i]`. Shouldn't the name of the original authors of the code also appear on there? (or, simpler: shouldn't it at least include the names that were present at the top of `env.ml[i]`?)

gasche commented:
Good point on the copyright; in the past I've just put my copyright when creating new files, but they were completely new.

Re. #2231: given that the present PR is non-urgent, I guess that the most reasonable thing is for me to wait until those recent PRs are settled to merge -- and deal with the rebase pains on my end. Grmbl...

trefis commented on typing/env.ml:
```diff
@@ -674,83 +552,37 @@ let rec print_address ppf = function
 
 (* The name of the compilation unit currently compiled.
    "" if outside a compilation unit. *)
+module Current_unit : sig
```
why not `Current_unit_name` then?

gasche replied:
done

trefis commented on toplevel/topdirs.ml:
```diff
@@ -118,15 +118,7 @@ let _ = add_directive "cd" (Directive_string dir_cd)
+  try Env.import_crcs ~source:filename cu.cu_imports
   with Consistbl.Inconsistency(name, user, auth) ->
     fprintf ppf "@[<hv 0>The files %s@ and %s@ \
                  disagree over interface %s@]@."
```
Why not put the `try .. with` in `Env.import_crcs` itself? Given that it's the one calling `Consistbl.check`.

Also, it looks like `Toploop` calls that function during initialization, without any `try ... with`; apparently it can't fail, but if it ever did, it would be with an uncaught exception, which doesn't seem very nice.

gasche replied:
I prefer for the refactoring commit to stay as close to the existing behavior as possible, and I don't know what Toploop is doing with this function, so I did not change anything here.

trefis commented on utils/misc.ml:
```diff
@@ -881,8 +881,80 @@ let print_if ppf flag printer arg =
 type alerts = string Stdlib.String.Map.t
+
+
+module EnvLazy = struct
```
Do we really want to grow misc? Couldn't `EnvLazy` be in its own file?

gasche replied:
I would rather grow misc than risking to introduce module name conflicts. We can always revisit this later.

gasche commented:
I just rebased on top of trunk. I'll re-read the rebased diffs, and my plan is to merge once the CI passes.

lpw25 commented:
Obviously very late to this, but I confirm that the CRC related change is correct.
